triggers_windows.go

```diff
@@ -1,55 +1,204 @@
+					"event":   eventName,
+					"message": "did not open event because process has already exited",
+				}))
+			} else {
+				grip.Error(message.WrapError(err, message.Fields{
+					"id":      info.ID,
+					"pid":     info.PID,
+					"source":  mongodShutdownSignalTriggerSource,
+					"event":   eventName,
+					"message": "failed to open event",
+				}))
+			}
 			return false
 		}
 		defer CloseHandle(event)
 
 		if err := SetEvent(event); err != nil {
-			grip.Errorf(errors.Wrapf(err, "failed to signal event '%s'", eventName).Error())
+			if mongodExited(err) {
+				grip.Debug(message.WrapError(err, message.Fields{
+					"id":      info.ID,
+					"pid":     info.PID,
+					"source":  mongodShutdownSignalTriggerSource,
+					"event":   eventName,
+					"message": "did not signal event because process has already exited",
+				}))
+			} else {
+				grip.Error(message.WrapError(err, message.Fields{
+					"id":      info.ID,
+					"pid":     info.PID,
+					"source":  mongodShutdownSignalTriggerSource,
+					"event":   eventName,
+					"message": "failed to signal event",
+				}))
+			}
 			return false
 		}
 
+		waitStatus, err := WaitForSingleObject(proc, mongodShutdownEventTimeout)
+		if err != nil || waitStatus != WAIT_OBJECT_0 {
+			grip.Error(message.WrapError(err, message.Fields{
+				"id":      info.ID,
+				"message": "failed to wait on process",
+				"source":  mongodShutdownSignalTriggerSource,
+			}))
+			return false
+		}
+		if waitStatus != WAIT_OBJECT_0 {
+			grip.Error(message.WrapError(getWaitStatusError(waitStatus), message.Fields{
+				"id":      info.ID,
+				"message": "wait did not return success",
+				"source":  mongodShutdownSignalTriggerSource,
+			}))
+			return false
+		}
+
+		return true
+	}
+}
+
+// makeCleanTerminationSignalTrigger terminates a process so that it will return exit code 0.
+func makeCleanTerminationSignalTrigger() SignalTrigger {
+	return func(info ProcessInfo, sig syscall.Signal) bool {
+		if sig != syscall.SIGTERM {
+			return false
+		}
+
+		proc, err := OpenProcess(PROCESS_TERMINATE|PROCESS_QUERY_INFORMATION, false, uint32(info.PID))
+		if err != nil {
+			// OpenProcess returns ERROR_INVALID_PARAMETER if the process has already exited.
+			if err == ERROR_INVALID_PARAMETER {
+				grip.Debug(message.WrapError(err, message.Fields{
+					"id":      info.ID,
+					"pid":     info.PID,
+					"source":  cleanTerminationSignalTriggerSource,
+					"message": "did not open process because it has already exited",
+				}))
+			} else {
+				grip.Error(message.WrapError(err, message.Fields{
+					"id":      info.ID,
+					"pid":     info.PID,
+					"source":  cleanTerminationSignalTriggerSource,
+					"message": "failed to open process",
+				}))
+			}
+			return false
+		}
+		defer CloseHandle(proc)
+
+		if err := TerminateProcess(proc, 0); err != nil {
+			// TerminateProcess returns ERROR_ACCESS_DENIED if the process has already died.
+			if err != ERROR_ACCESS_DENIED {
```

Comment:
Should there be a branch where `err == ERROR_ACCESS_DENIED`? I don't follow the `err != ERROR_ACCESS_DENIED` logic here.

Reply:
This block (L170-178) is for the case when the process is known to be dead already.

If `err == ERROR_ACCESS_DENIED` the logic after this block (L180-200) has to verify that the process actually exited. This is done by checking the exit code, and if it returns `STILL_ACTIVE`, the process is still alive.